File: 1_het/mini_prog/RagApi/Services/ChatService.cs
using System.Net.Http.Json;
using System.Text.Json;
using ef;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
public class ChatService
{
    private readonly CodeDbContext db;
    private readonly HttpClient httpClient;
    private readonly IConfiguration configuration;
    public ChatService(CodeDbContext cdb, HttpClient client, IConfiguration conf)
    {
        db = cdb;
        httpClient = client;
        configuration = conf;
    }
    public async Task<ServiceResult> AddConversationAsync(ConversationData data)
    {

        var exist = await db.Projects.AnyAsync(x => x.ID == data.projectId);
        if (!exist) return ServiceResult.Fail("Project doesn't exist");

        var conversation = new DbConversation
        {
            Title = data.title,
            ProjectID = data.projectId,
            CreatedAtUtc = DateTime.UtcNow,
            UpdatedAtUtc = DateTime.UtcNow
        };
        db.Conversations.Add(conversation);
        await db.SaveChangesAsync();
        int id = conversation.ID;
        return ServiceResult.Success(id);
        //Id return????????????????????????????????,

    }
    public async Task<ServiceResult> LoadConversationsAsync(ProjectID id)
    {
        var exist = await db.Projects.AnyAsync(x => x.ID == id.id);
        if (!exist) return ServiceResult.Fail("Project doesn't exist");

        var conv = await db.Conversations
        .Where(c => c.ProjectID == id.id)
        .Include(c => c.Messages)
        .Select(x => new Conversation
        {
            ID = x.ID,
            Title = x.Title,
            CreatedAtUtc = x.CreatedAtUtc,
            UpdatedAtUtc = x.UpdatedAtUtc,
            Messages = x.Messages
            .Select(m => new Message
            {
                ID = m.ID,
                Role = m.Role,
                Content = m.Content,
                CreatedAtUtc = m.CreatedAtUtc
            })
            .ToList()
        }).ToListAsync();

        return ServiceResult.Success(conv);
        /////nem jó az include még

    }
    public async Task<ServiceResult> AddMessageAsync(MessageData data)
    {
        var exist = await db.Conversations.AnyAsync(x => x.ID == data.convId);
        if (!exist) return ServiceResult.Fail("The conversation doesn't exist.");

        var message = new DbMessage
        {
            ConversationID = data.convId,
            Role = data.role,
            CreatedAtUtc = DateTime.UtcNow,
            Content = data.content
        };
        db.Messages.Add(message);
        await db.SaveChangesAsync();
        return ServiceResult.Success();
    }

    public async Task<ServiceResult> LoadMessagesAsync(Id data)
    {
        var exist = await db.Conversations.AnyAsync(x => x.ID == data.id);
        if (!exist) return ServiceResult.Fail("The conversation doesn't exist.");

        var messages = await db.Messages
        .Where(x => x.ConversationID == data.id)
        .Select(m => new Message
        {
            ID = m.ID,
            Role = m.Role,
            CreatedAtUtc = m.CreatedAtUtc,
            Content = m.Content
        }).ToListAsync();

        return ServiceResult.Success(messages);
    }
    private async Task<float[]> Embed(HttpClient http, string text)
    {

        var payload = new { model = configuration["AI:EmbedModel"], prompt = text };
        
        var response = await http.PostAsJsonAsync($"{configuration["AI:OllamaUrl"]}/api/embeddings", payload);
        response.EnsureSuccessStatusCode();
    
        using var doc = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
        return doc.RootElement.GetProperty("embedding")
           .EnumerateArray()
           .Select(x => x.GetSingle())
           .ToArray();
    }
    public async Task<ServiceResult> SendMessageAsync(ChatRequest prompt)
    {
        Console.WriteLine(configuration["AI:EmbedModel"]);
        Console.WriteLine(configuration["AI:OllamaUrl"]);
        httpClient.Timeout = TimeSpan.FromMinutes(5);
        try
        {
            System.Console.WriteLine("Még jóóóóóóóóóó   000000");
            //text -> embedding
            var vec = await Embed(httpClient, prompt.prompt);

            System.Console.WriteLine(vec.Length);
            //Embedded text -> qdrant: search
            var searchPayload = new
            {
                vector = vec,
                limit = 3,
                with_payload = true

            };
            System.Console.WriteLine("Még jóóóóóóóóóó   000001");
            var response = await httpClient.PostAsJsonAsync($"{configuration["AI:QdrantUrl"]}/collections/{configuration["AI:Collection"]}/points/search", searchPayload);
            
            var content = await response.Content.ReadAsStringAsync();

            Console.WriteLine($"Status: {response.StatusCode}");
            Console.WriteLine(content);
            response.EnsureSuccessStatusCode();
            System.Console.WriteLine("Még jóóóóóóóóóó   000001");
            //
            using var searchDoc = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
            var results = searchDoc.RootElement.GetProperty("result").EnumerateArray().ToList();
            System.Console.WriteLine("még jóóóó 1");
            //Have results or not
            if (results.Count == 0)
            {
                return ServiceResult.Success(new { answer = "Nem találom a dokumentumokban." });
            }

            //making context
            var contextParts = new List<string>();
            foreach (var r in results)
            {
                var payload = r.GetProperty("payload");
                var text = payload.GetProperty("text").GetString() ?? " ";
                var docName = payload.GetProperty("docName").GetString() ?? "doc";
                var idx = payload.GetProperty("chunkText").GetInt32();
                contextParts.Add($"[Forrás: {docName} #{idx}]\n{text}");
            }
            var context = string.Join("\n\n---\n\n", contextParts);
            System.Console.WriteLine("még jóóóó 2");
            // prompt to model
            var finalPrompt = $"""
            Te egy asszisztens vagy, aki KIZÁRÓLAG az alábbi KONTEKSZTUS alapján válaszol.Mindig azon a nyelven válaszolj, amilyen nyelven a kérdés elhangzott. A szavak közé tegyél szóközöket, ha kell, úgy higy értelmesen legyenek elválasztva.Ha a válasz nem található a kontextusban, mondd: "Nem találom a dokumentumokban."

            KONTEKSZTUS:
            {context}

            KÉRDÉS:
            {prompt.prompt}
            """;

            //Ollama generate
            var genPayload = new { model = configuration["AI:GenModel"], prompt = finalPrompt, stream = false };
            var genAnswer = await httpClient.PostAsJsonAsync($"{configuration["AI:OllamaUrl"]}/api/generate", genPayload);
            genAnswer.EnsureSuccessStatusCode();

            using var genDoc = JsonDocument.Parse(await genAnswer.Content.ReadAsStringAsync());

            var answer = genDoc.RootElement.GetProperty("response").GetString() ?? " ";

            Console.WriteLine("A válasz: " + answer);
            return ServiceResult.Success(new { answer });
        }
        catch (Exception e)
        {
            return ServiceResult.Fail(e.Message);
        }

    }

}

File: 1_het/mini_prog/AI/app/services/chat_service.py
import httpx

from app.models.models import ServiceResult, ChatRequest
from app.services.file_service import FileService
from app.config import QDRANT_URL, QDRANT_COLLECTION, GEN_MODEL, OLLAMA_URL
from app.services.rag_graph import rag_graph
#langgraph nálküli válaszgenerálás
async def send_message_async(prompt: ChatRequest) -> ServiceResult:
    async with httpx.AsyncClient(timeout=10000) as http_client:
        # text -> embedding
        #JAVITAS
        file_service=FileService()
        vec = await file_service.embed(http_client, prompt.prompt)
        
        results_sum=await search_vector("summary",vec,http_client)
        results_code=await search_vector("code",vec,http_client)
        results_text=await search_vector("text",vec,http_client)
        print(f"code: {len(results_code)}")
        print(f"sum: {len(results_sum)}")
        print(f"text: {len(results_text)}")
        #print("még jóóóó 1")

        if len(results_sum)+len(results_code)+len(results_text) == 0:
            return ServiceResult.success({
                "answer": "Nem találom a dokumentumokban."
            })

        # context összeállítása
        all_results = results_sum + results_code + results_text
        
        for r in all_results:
            print(r.get("score", 0))

        # score alapján csökkenő sorrend, bizonyos score felett rakom cak bele JAVITAS
        top_results = sorted(
            [r for r in all_results if r.get("score", 0) >= 0.35],
            #all_results,
            key=lambda r: r.get("score", 0),
            reverse=True
        )[:5]
        print("TOP EREDMÉNYEK SZÁMA:")
        print(len(top_results))
        if not top_results:
            return ServiceResult.success({
                "answer": "Nem találom a dokumentumokban."
            })
        context_parts = []
        create_context(top_results,context_parts)
        #create_context(results_code,context_parts)
        
        #create_context(results_sum,context_parts)
        #create_context(results_text,context_parts)

        context = "\n\n---\n\n".join(context_parts)
        print("CONTEXT")
        print(context)
        print("context vége")
        #print("még jóóóó 2")

        # prompt
        final_prompt = f"""Te egy asszisztens vagy, aki KIZÁRÓLAG az alábbi KONTEKSZTUS alapján válaszol.
        Mindig azon a nyelven válaszolj, amilyen nyelven a kérdés elhangzott.
        A szavak közé tegyél szóközöket, ha kell, úgy hogy értelmesen legyenek elválasztva.
        Ha a válasz nem található a kontextusban, mondd: "Nem találom a dokumentumokban."

        KONTEKSZTUS:
        {context}

        KÉRDÉS:
        {prompt}
        """

        # Ollama generate
        gen_payload = {
            "model": GEN_MODEL,
            "prompt": final_prompt,
            "stream": False
        }

        gen_response = await http_client.post(
            f"{OLLAMA_URL}/api/generate",
            json=gen_payload
        )
        gen_response.raise_for_status()

        gen_data = gen_response.json()
        answer = gen_data.get("response") or " "

        print("A válasz:", answer)

        return ServiceResult.success({
            "answer": answer
        })
async def send_message_async_langgraph(prompt: ChatRequest) -> ServiceResult:
    result = await rag_graph.ainvoke({
        "question": prompt.prompt
    })

    return ServiceResult.success({
        "answer": result.get("answer", "Nem találom a dokumentumokban.")
    })

'''   
async def create_context(results, context_parts):
        # context összeállítása

        for r in results:
            payload = r.get("payload", {})

            text = payload.get("text") or " "
            doc_name = payload.get("docName") or "doc"
            idx = payload.get("chunkText", 0)

            context_parts.append(f"[Forrás: {doc_name} #{idx}]\n{text}")
'''  